node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -1042,7 +1012,17 @@ pub(crate) async fn run<Context>(
 
 					send_collation(&mut state, next, receipt, pov).await;
 				}
-			}
+			},
+			_ = reconnect_stream.next() => {
+				let now = Instant::now();
+				if state
+					.last_connected_at
+					.map_or(false, |timestamp| now - timestamp > RECONNECT_TIMEOUT)
```

Comment:
Why don't we just wait for `RECONNECT_TIMEOUT`? Instead of checking whether it elapsed in constant intervals?

Reply:
We ensure that we reconnect at least once every `RECONNECT_TIMEOUT` seconds

For example, if we just advertised a collation, there's no need to issue another reconnect request.

Reply:
So we override the `Delay` future each time we receive an advertisment?

Reply:
We overwrite the timestamp on every call to `connect_to_validators`, yes
https://github.com/paritytech/polkadot/blob/ab57c1e02eb33b4b37ae8269c7d5f36eab4daa6b/node/network/collator-protocol/src/collator_side/mod.rs#L378-L379

Reply:
Ok phrased differently: Why that inifinte stream of a delay shorter than what we actually need to wait for. Instead we could just have a `Delay` future that wakes exactly when needed by getting overridden at that very line?

Reply:
Probably not much difference, instead we wold juggle around pinned futures which doesn't look like a clean approach to me. We do some timeout checks on both sides of protocol so I decided to keep it consistent

Reply:
Not sure I follow, but it also isn't that important right now.